h2o-core/src/main/java/water/FrameSizeMonitor.java

```diff
@@ -0,0 +1,211 @@
+    private static final String SAFE_FREE_MEM_DEFAULT_COEF = "0.2";
+
+    private static final int LOG_LEVEL;
+    private static final boolean ENABLED;
+    private static final float SAFE_FREE_MEM_COEF;
+    private static final int SLEEP_MS = 100;
+    private static final int MB = 1024 * 1024;
+    
+    private static final Map<Key<Job>, FrameSizeMonitor> registry = new HashMap<>();
+    
+    static {
+        LOG_LEVEL = Integer.parseInt(System.getProperty(LOG_LEVEL_PROP, String.valueOf(Log.DEBUG)));
+        ENABLED = H2O.getSysBoolProperty(ENABLED_PROP, false);
+        SAFE_FREE_MEM_COEF = Float.parseFloat(H2O.getSysProperty(SAFE_COEF_PROP, SAFE_FREE_MEM_DEFAULT_COEF));
+    }
+
+    private final Key<Job> jobKey;
+    private final Set<FVecParseWriter> writers = new HashSet<>();
+    private final long totalMemory = getTotalMemory();
+
+    private long committedMemory = 0;
+    
+    FrameSizeMonitor(Key<Job> jobKey) {
+        this.jobKey = jobKey;
+    }
+    
+    public static void get(Key<Job> jobKey, Consumer<FrameSizeMonitor> c) {
+        synchronized (registry) {
+            if (registry.containsKey(jobKey)) {
+                c.accept(registry.get(jobKey));
+            } else if (ENABLED) {
+                if (jobKey.get().stop_requested()) {
+                    // throw an exception to stop the parsing
+                    throw new IllegalStateException("Memory is running low. Forcefully terminating.");
+                } else {
+                    FrameSizeMonitor monitor = new FrameSizeMonitor(jobKey);
+                    registry.put(jobKey, monitor);
+                    H2O.submitTask(new LocalMR<>(monitor, 1));
+                    c.accept(monitor);
+                }
+            }
+        }
+    }
+    
+    private static void finish(Key<Job> jobKey) {
+        synchronized (registry) {
+            registry.remove(jobKey);
+        }
+    }
+
+    @Override
+    protected void map(int id) {
+        float nextProgress = 0.02f;
+        Job<Frame> job = jobKey.get();
+        while (job.isRunning() && nextProgress < 1f) {
+            if (!MemoryManager.canAlloc()) {
+                Log.info("FrameSizeMonitor: MemoryManager is running low on memory, stopping job " + jobKey + " writing frame " + job._result);
+                job.fail(new RuntimeException("Aborting due to critically low memory."));
+                break;
+            }
+            float currentProgress = job.progress();
+            if (currentProgress >= nextProgress) {
+                if (isMemoryUsageOverLimit() && isFrameSizeOverLimit(currentProgress, job)) {
+                    job.fail(new RuntimeException("Aborting due to projected memory usage too high."));
+                    break;
+                } else if (nextProgress < 0.1f) {
+                    nextProgress = currentProgress + 0.01f;
+                } else {
+                    nextProgress = currentProgress + 0.1f;
```

Comment:
I can imagine making this interval smaller for large datasets, checking per every 10 percent parsed could go wrong. Based on how much data was parsed in the first 10 percent of progress maybe ? If the data initially parsed take up more than 10% of available memory, it may be reasonable to make this interval shorter.

Reply:
I agree this approach has its weaknesses, but there is a balance of checking too often and not checking often enough. This code is not meant to be bullet-proof, but has proven by my testing to work quite well.
Concrete suggestions on how to improve this are welcome.